File: src/dart_py/detector.py
"""OpenCV green-light detector for CanMV K230."""
import cv2
from ulab import numpy as np


class Detector:
    def __init__(
        self,
        hsv_low=(45, 80, 120),
        hsv_high=(85, 255, 255),
        min_area=3.0,
        max_area=5000.0,
        min_circularity=0.2,
        max_aspect_ratio_diff=0.5,
    ):
        self.hsv_low = _np_u8(hsv_low)
        self.hsv_high = _np_u8(hsv_high)
        self.min_area = float(min_area)
        self.max_area = float(max_area)
        self.min_circularity = float(min_circularity)
        self.max_aspect_ratio_diff = float(max_aspect_ratio_diff)

    def detect(self, image):
        """Detect the target in a CanMV image.Image or RGB888 ndarray."""
        image_np = _as_ndarray(image)
        if image_np is None:
            return _lost_result()

        hsv_image = cv2.cvtColor(image_np, cv2.COLOR_RGB2HSV)
        mask = cv2.inRange(hsv_image, self.hsv_low, self.hsv_high)

        contours, _ = cv2.findContours(
            mask,
            cv2.RETR_EXTERNAL,
            cv2.CHAIN_APPROX_SIMPLE,
        )

        best_contour = None
        best_stats = None
        best_score = 0.0
        for contour in contours:
            stats = self._candidate_stats(contour)
            if stats is None:
                continue
            score = stats["area"]
            if score > best_score:
                best_score = score
                best_contour = contour
                best_stats = stats

        if best_contour is None:
            return _lost_result()
        return _result_from_contour(best_contour, best_stats)

    def _candidate_stats(self, contour):
        area = cv2.contourArea(contour)
        if area < self.min_area:
            return None
        if self.max_area > 0 and area > self.max_area:
            return None

        perimeter = cv2.arcLength(contour, True)
        if perimeter <= 0:
            return None

        circularity = 4 * np.pi * area / (perimeter * perimeter)
        if circularity < self.min_circularity:
            return None

        x, y, w, h = cv2.boundingRect(contour)
        if w <= 0 or h <= 0:
            return None

        aspect_ratio_diff = abs(w - h) / max(w, h)
        if aspect_ratio_diff > self.max_aspect_ratio_diff:
            return None

        return {
            "area": float(area),
            "bbox": (x, y, w, h),
            "circularity": float(circularity),
        }


def _as_ndarray(image):
    if hasattr(image, "to_numpy_ref"):
        return image.to_numpy_ref()
    return image


def _np_u8(values):
    return np.array(values, dtype=np.uint8)


def _result_from_contour(contour, stats):
    x, y, w, h = stats["bbox"]
    moments = cv2.moments(contour)
    if moments["m00"] != 0:
        cx = float(moments["m10"] / moments["m00"])
        cy = float(moments["m01"] / moments["m00"])
    else:
        cx = float(x + 0.5 * w)
        cy = float(y + 0.5 * h)

    return {
        "detected": True,
        "x": cx,
        "y": cy,
        "bbox": (x, y, w, h),
        "area": float(stats["area"]),
        "circularity": float(stats["circularity"]),
    }


def _lost_result():
    return {
        "detected": False,
        "x": -1.0,
        "y": -1.0,
        "bbox": None,
        "area": 0.0,
        "circularity": 0.0,
    }

File: src/dart_cpp/src/detector.cpp
#include <detector.hpp>
#include <stdexcept>

void Detector::initialize(
    const cv::Scalar& hsv_low,
    const cv::Scalar& hsv_high,
    double min_circ,
    double min_a,
    double max_a,
    double max_aspect_diff)
{
    hsv_threshold_low = hsv_low;
    hsv_threshold_high = hsv_high;
    min_circularity = min_circ;
    min_area = min_a;
    max_area = max_a;
    max_aspect_ratio_diff = max_aspect_diff;

    initialized = true;
}

std::optional<GreenTarget> Detector::detect(const cv::Mat& image)
{
    if (!initialized)
    {
        throw std::runtime_error("Detector not initialized. Call initialize() before detect().");
    }

    if (image.empty())
    {
        return std::nullopt;
    }

    // Convert the image to HSV color space
    cv::Mat hsv_image;
    cv::cvtColor(image, hsv_image, cv::COLOR_BGR2HSV);

    // Threshold the HSV image to get only green colors
    cv::Mat mask;
    cv::inRange(hsv_image, hsv_threshold_low, hsv_threshold_high, mask);

    // 使用闭运算填补目标区域中的小孔洞和断裂：先膨胀，再腐蚀
    cv::Mat morphology_kernel = cv::getStructuringElement(
        cv::MORPH_ELLIPSE, cv::Size(5, 5));
    cv::morphologyEx(mask, mask, cv::MORPH_CLOSE, morphology_kernel);

    // Find contours in the mask
    std::vector<std::vector<cv::Point>> contours;
    cv::findContours(mask, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

    std::optional<GreenTarget> best_target;

    for (const auto& contour : contours)
    {
        double area = cv::contourArea(contour);
        if (area < min_area || (max_area > 0.0 && area > max_area))
        {
            continue;
        }

        double perimeter = cv::arcLength(contour, true);
        if (perimeter <= 0.0)
        {
            continue;
        }

        // 过滤过于细长的轮廓，避免将非目标绿色区域识别为目标
        cv::Rect bounding_box = cv::boundingRect(contour);
        if (bounding_box.width <= 0 || bounding_box.height <= 0)
        {
            continue;
        }

        double aspect_ratio_diff =
            static_cast<double>(std::abs(bounding_box.width - bounding_box.height)) /
            static_cast<double>(std::max(bounding_box.width, bounding_box.height));
        if (aspect_ratio_diff > max_aspect_ratio_diff)
        {
            continue;
        }

        // 计算轮廓圆形度
        double circularity = 4 * CV_PI * area / (perimeter * perimeter);
        if (circularity < min_circularity)
        {
            continue;
        }

        // 使用最小包围圆计算目标中心和半径
        cv::Point2f center;
        float radius;
        cv::minEnclosingCircle(contour, center, radius);

        GreenTarget candidate(center, radius, bounding_box, area, circularity);
        if (!best_target || candidate.area > best_target->area)
        {
            best_target = candidate;
        }
    }

    return best_target;
}
